### PR TITLE
fix(runtime): make child workflow replay idempotent

### LINKED ISSUE #1200: fix(runtime): enforce atomic workflow transition boundaries
## Problem

Some workflow runtime paths still perform durable side effects before the runtime job completion is committed under the lease-owned boundary. That creates crash/retry windows where a child workflow or submission can be persisted even when the parent job completion is later ignored or retried.

## Evidence

- `crates/harness-workflow/src/runtime/worker.rs:96` executes the activity first, records `ActivityResultReady`, then commits completion at `commit_runtime_activity_completion_if_owned`.
- `crates/harness-server/src/workflow_runtime_worker/child_workflow.rs:57` upserts child workflow definitions and instances before the parent completion boundary.
- `crates/harness-server/src/workflow_runtime_worker/child_workflow.rs:119` can auto-submit the issue through `record_issue_submission` before the parent job completion is committed.
- `crates/harness-workflow/src/runtime/store.rs:268` already has `apply_decision_transition`, but not every transition path uses one transaction boundary.
- `crates/harness-workflow/src/runtime/store.rs:840` and `:2044` dedupe commands by `(workflow_id, dedupe_key)` while preserving old payload fields on conflict.

## Impact

A worker crash, lease loss, or retry can leave parent and child workflows out of sync, duplicate submissions, or stale command payloads. Operators then debug symptoms in issue workflows instead of the real runtime ownership boundary.

## Acceptance criteria

- Define one transaction boundary for workflow instance update, event append, decision record, command outbox writes, and runtime job completion.
- Move child workflow start/submission effects behind that boundary or make them idempotent with explicit parent-job ownership checks.
- Make command dedupe conflicts preserve or update the expected payload fields intentionally, with tests for conflict behavior.
- Add crash/retry tests that simulate a side effect before completion and prove replay cannot duplicate or drift parent/child state.
- Document the invariant in the workflow runtime hardening design.

## Related

A narrower PR-feedback atomicity issue was tracked in #1144. This issue covers the remaining architectural boundary, not only that one closed path.

### PR DESCRIPTION
**Pull request by majiayu000**
## Summary

- add a lease-ownership guard before built-in workflow lifecycle side effects
- make GitHub issue child workflow start and auto-submit replay idempotent for the same parent runtime command
- centralize workflow command insert/upsert conflict handling and refresh pending command payload metadata intentionally
- document the child workflow replay invariant in the runtime hardening design

## Validation

- cargo fmt --all -- --check
- cargo test -p harness-workflow runtime_store_pending_dedupe_refreshes_command_payload
- cargo test -p harness-server runtime_job_worker_replays_auto_submit_without_duplicate_child_side_effects
- cargo check -p harness-workflow --all-targets
- cargo check -p harness-server --all-targets
- cargo test --workspace
- cargo clippy --workspace --all-targets -- -D warnings

Closes #1200

### CODE REVIEW

File: crates/harness-workflow/src/runtime/store.rs [hunk 1]
```diff
@@ -1681,6 +1636,16 @@ impl WorkflowRuntimeStore {
...
     }
 
+    pub async fn runtime_job_matches_running_lease(
+        &self,
+        expected: &RuntimeJob,
+    ) -> anyhow::Result<bool> {
+        let Some(current) = self.get_runtime_job(&expected.id).await? else {
+            return Ok(false);
+        };
+        Ok(current.status == RuntimeJobStatus::Running && current.lease == expected.lease)
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-workflow/src/runtime/store.rs]: Addressed in 79441c8e. The runtime job now carries a lease_generation that increments on claim/reclaim and is preserved on renewal, so built-in lifecycle side effects accept renewed same-worker leases while rejecting stale same-owner reclaims. Added regression coverage for both cases.

File: crates/harness-server/src/workflow_runtime_worker/child_workflow.rs [hunk 2]
```diff
@@ -117,30 +182,32 @@ pub(super) async fn execute_start_child_workflow(
...
-                dependencies_blocked: !depends_on.is_empty(),
-                source: Some(source.as_str()),
-                external_id: Some(external_id.as_str()),
-            },
-        )
-        .await?;
-        child_submission = Some(submission);
-        if let Some(updated) = store.get_instance(&child.id).await? {
-            child = updated;
+        if !issue_submission_recorded(&child, &task_id) {
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-server/src/workflow_runtime_worker/child_workflow.rs]: Addressed in 23d97c89: the auto-submit replay guard now checks both child data and existing IssueSubmitted events for the task id before calling record_issue_submission, and the replay regression removes the data markers to prove the event path prevents duplicate events and decisions.

File: crates/harness-server/src/workflow_runtime_worker/child_workflow.rs [hunk 3]
```diff
@@ -117,30 +209,32 @@ pub(super) async fn execute_start_child_workflow(
...
-                dependencies_blocked: !depends_on.is_empty(),
-                source: Some(source.as_str()),
-                external_id: Some(external_id.as_str()),
-            },
-        )
-        .await?;
-        child_submission = Some(submission);
-        if let Some(updated) = store.get_instance(&child.id).await? {
-            child = updated;
+        if !issue_submission_recorded(store, &child, &task_id).await? {
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-server/src/workflow_runtime_worker/child_workflow.rs:133]: Addressed in 0f7f6a23: the replay guard now requires completed submission side effects before skipping resubmission, and record_issue_submission reuses the existing IssueSubmitted event to finish missing decision, command, and child state writes. Added an event-only partial-write regression.

File: crates/harness-server/src/workflow_runtime_worker/child_workflow.rs [hunk 4]
```diff
@@ -27,6 +144,7 @@ pub(super) async fn execute_start_child_workflow(
     let Some(store) = state.core.workflow_runtime_store.as_ref() else {
         anyhow::bail!("workflow runtime store is unavailable");
     };
+    ensure_runtime_job_still_owns_lease(store, job).await?;
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-server/src/workflow_runtime_worker/child_workflow.rs:35]: Addressed in ba2091e4. Prompt-task submissions now reuse existing PromptSubmitted events/decisions on replay, and prompt, quality-gate, and PR-feedback child-start/request paths now use command-scoped replay guards instead of appending duplicate ChildWorkflowStarted/request events. Added runtime_worker_non_issue_replay_tests for all three non-issue child workflow branches.

File: crates/harness-server/src/workflow_runtime_submission.rs [hunk 5]
```diff
@@ -282,7 +278,112 @@ async fn apply_decision(
...
     })
 }
-
+async fn issue_submission_event_id_or_append(
+    store: &WorkflowRuntimeStore,
+    workflow_id: &str,
+    ctx: &IssueSubmissionRuntimeContext<'_>,
+) -> anyhow::Result<String> {
+    if let Some(event_id) = issue_submission_event_id(store, workflow_id, ctx.task_id).await? {
+        return Ok(event_id);
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-server/src/workflow_runtime_submission.rs]: Addressed in a1d01a9c. Submission event reuse is now limited to replayable side effects: missing decisions/commands, unapplied state, or active pending/dispatching commands. Completed historical attempts append a new submission event, and repeat issue/prompt submissions with the same task id receive event-scoped command dedupe keys so the reopened attempt gets fresh pending work. Added regressions for issue and prompt resubmission after completed commands.

File: crates/harness-server/src/workflow_runtime_worker/child_workflow.rs [hunk 6]
```diff
@@ -117,30 +130,32 @@ pub(super) async fn execute_start_child_workflow(
...
-                dependencies_blocked: !depends_on.is_empty(),
-                source: Some(source.as_str()),
-                external_id: Some(external_id.as_str()),
-            },
-        )
-        .await?;
-        child_submission = Some(submission);
-        if let Some(updated) = store.get_instance(&child.id).await? {
-            child = updated;
+        if !issue_submission_recorded(store, &child, &task_id).await? {
```
[the remark this answers is not in the document]
Comment (majiayu000) [reply crates/harness-server/src/workflow_runtime_worker/child_workflow.rs:133]: Addressed in 4b2d11ac. The auto-submit preflight no longer treats historical `submission_id` / `task_id` / `task_ids` data as a completed replay marker; it now relies on completed side-effect evidence, allowing `record_issue_submission` to create a fresh event-scoped command for reopened issues. Added `runtime_job_worker_auto_submit_reopens_after_completed_historical_task_id_data` to cover this regression.
